src/backend/parser/parse_agg.c

```diff
@@ -1505,16 +1537,140 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
 		List *exprs = (List *)grpcl;
 		ListCell *lc;
 
-		foreach (lc, exprs)
+		foreach(lc, exprs)
 		{
-			result = list_concat(result, get_groupclause_exprs((Node *)lfirst(lc),
+			result = list_concat(result, get_groupclause_tles((Node *)lfirst(lc),
 															   targetList));
 		}
 	}
 
 	return result;
 }
 
+/*
+ * get_com_grouping_ressortgroupref_routine -
+ *     Return a list of common ressortgrouprefs of the current grouping
+ *     expression recursively. It is assumed that attributes used in all
+ *     groupings are already present in the current grouping. Otherwise,
+ *     missing references are removed from the com_refs list.
+ */
+static List*
+get_com_grouping_ressortgroupref_routine(Node *grpcl, List *targetList, List *com_refs)
+{
+	List *result = NIL;
+
+	if ( !grpcl )
+		return result;
+
+	Assert(IsA(grpcl, SortGroupClause) ||
+		   IsA(grpcl, GroupingClause) ||
+		   IsA(grpcl, List));
+
+	if (IsA(grpcl, SortGroupClause))
+	{
+		TargetEntry *tle = get_sortgroupclause_tle((SortGroupClause *) grpcl, targetList);
+		result = lappend_int(result, tle->ressortgroupref);
+	}
```

Comment:
If you add `return result;` here, you may remove the last `else` with brackets and one tab in each line inside.

Reply:
Done. This solution seems less clear.